### DISCUSSION TITLE
Brain teaser: nested data structure traversal

### DISCUSSION THREAD

**Discussion by myitcv** [cue-lang/cue#2158] [Q&A]
_This brain teaser is a bit of an experiment as part of the [CUE Community Call 2022-11-29](https://github.com/cue-lang/cue/discussions/2152)_

The [Benthos](https://www.benthos.dev/) project provides an experimental [CUE configuration interface](https://www.benthos.dev/docs/configuration/using_cue). 

Another key aspect of Benthos is data transformation/mapping.

In conversations with @disintegrator and @jeffail we asked for gnarly problems they have encountered in Benthos to see how CUE stacks up.

@Jeffail shared the following https://github.com/Jeffail/nested-data-structure-traversal.

For convenience, I have quoted the problem statement below. 

The goal is to write a solution using CUE, and post your solution as a reply. If you want to avoid spoilers, be careful not to scroll too far down! 

> ## The problem
> 
> The algorithm should receive a list of sections. A section is a key-value data structure, with a "title", a "reset_lesson_position" boolean, and a list of "lessons". A lesson is a key-value data structure with the "name" field.
> 
> Your job is to traverse the list of sections, adding a position (starting from 1) to each section, and traverse the list of lessons adding a position (starting from 1) to each lesson. Note, however, the lessons position is shared across sections. The lesson position should also be reset if "reset_lesson_position" is true.
> 
> Here is an example input (formatted in JSON for convenience):
> 
> ```json
> [
>   {
>     "title": "Getting started",
>     "reset_lesson_position": false,
>     "lessons": [
>       {"name": "Welcome"},
>       {"name": "Installation"}
>     ]
>   },
> 
>   {
>     "title": "Basic operator",
>     "reset_lesson_position": false,
>     "lessons": [
>       {"name": "Addition / Subtraction"},
>       {"name": "Multiplication / Division"}
>     ]
>   },
> 
>   {
>     "title": "Advanced topics",
>     "reset_lesson_position": true,
>     "lessons": [
>       {"name": "Mutability"},
>       {"name": "Immutability"}
>     ]
>   }
> ]
> ```
> 
> The output should be (formatted in JSON for convenience):
> 
> ```json
> [
>   {
>     "title": "Getting started",
>     "reset_lesson_position": false,
>     "position": 1,
>     "lessons": [
>       {"name": "Welcome", "position": 1},
>       {"name": "Installation", "position": 2},
>     ]
>   },
> 
>   {
>     "title": "Basic operator",
>     "reset_lesson_position": false,
>     "position": 2,
>     "lessons": [
>       {"name": "Addition / Subtraction", "position": 3},
>       {"name": "Multiplication / Division", "position": 4}
>     ]
>   },
> 
>   {
>     "title": "Advanced topics",
>     "reset_lesson_position": true,
>     "position": 3,
>     "lessons": [
>       {"name": "Mutability", "position": 1},
>       {"name": "Immutability", "position": 2}
>     ]
>   }
> ]
> ```
>

**Comment by eonpatapon** [comment]
Here's mine ;)
Not very pretty!

```cue
package x

#Section: {
	title:                 string
	reset_lesson_position: bool
	lessons: [...#Lesson]
	position: int
}

#Lesson: {
	name:     string
	position: int
}

input: [...#Section]
input: [
	{
		"title":                 "Getting started"
		"reset_lesson_position": false
		"lessons": [
			{"name": "Welcome"},
			{"name": "Installation"},
		]
	},

	{
		"title":                 "Basic operator"
		"reset_lesson_position": false
		"lessons": [
			{"name": "Addition / Subtraction"},
			{"name": "Multiplication / Division"},
		]
	},

	{
		"title":                 "Advanced topics"
		"reset_lesson_position": true
		"lessons": [
			{"name": "Mutability"},
			{"name": "Immutability"},
		]
	},
]

output: [...#Section]
output: [
	for sidx, s in input {
		s & {
			reset_lesson_position: bool
			position:              sidx + 1
			lessons: [
				for lidx, l in s.lessons {
					{
						if reset_lesson_position {
							position: lidx + 1
						}

						if !reset_lesson_position {
							// Default case for first section
							position: *1 | int

							// We have a previous section, calculate position
							// by finding the last lesson position of the previous
							// section
							if lidx == 0 && sidx > 0 {
								_psection: output[sidx-1]
								position:  _psection.lessons[len(_psection.lessons)-1].position + 1
							}

							// Not the first lesson, calculate with the
							// previous lesson position
							if lidx > 0 {
								position: lessons[lidx-1].position + 1
							}
						}
					}
				},
			]

		}
	},
]
```

` cue export nested-data-structure-traversal.cue  -e output`
```json
[
    {
        "title": "Getting started",
        "reset_lesson_position": false,
        "lessons": [
            {
                "name": "Welcome",
                "position": 1
            },
            {
                "name": "Installation",
                "position": 2
            }
        ],
        "position": 1
    },
    {
        "title": "Basic operator",
        "reset_lesson_position": false,
        "lessons": [
            {
                "name": "Addition / Subtraction",
                "position": 3
            },
            {
                "name": "Multiplication / Division",
                "position": 4
            }
        ],
        "position": 2
    },
    {
        "title": "Advanced topics",
        "reset_lesson_position": true,
        "lessons": [
            {
                "name": "Mutability",
                "position": 1
            },
            {
                "name": "Immutability",
                "position": 2
            }
        ],
        "position": 3
    }
]
```

**Comment by verdverm** [comment]
```cue
answer: [
    for d, D in _data {
      // D

      // section position, always increasing
      position: d + 1
      
      // calculate the base value for lesson pos in this section, this is the interesting part
      _lp: int
      // set to zero for first and on reset
      if d == 0 || D.reset_lesson_position {
       _lp: 0
      }
      // set current to previous base + lesson count of previous section
      if d > 0 && D.reset_lesson_position == false {
        let prev = answer[d-1]
        _lp: prev._lp + len(prev.lessons)
      }
      
      // comprehend lessons and set position
      "lessons": [ for l, L in D.lessons {
        // L
        position: _lp + l + 1
      }]
    }
]
```

https://cuelang.org/play/?id=AyBgqY5xXZs#cue@export@cue
